src/main/kotlin/app/revanced/patcher/Patcher.kt

```diff
@@ -191,201 +61,137 @@ class Patcher(private val options: PatcherOptions) {
-                    decoder.currentPackage = ResPackage(resourceTable, 0, null)
-
-                    // create xml parser with the decoder
-                    val axmlParser = AXmlResourceParser()
-                    axmlParser.attrDecoder = decoder
-
-                    // parse package information with the decoder and parser which will set required values in the resource table
-                    // instead of decodeManifest another more low level solution can be created to make it faster/better
-                    XmlPullStreamDecoder(
-                        axmlParser, AndrolibResources().resXmlSerializer
-                    ).decodeManifest(
-                        extInputFile.directory.getFileInput("AndroidManifest.xml"), nullOutputStream
-                    )
-                }
-            }
-
-            // read of the resourceTable which is created by reading the manifest file
-            context.packageMetadata.let { metadata ->
-                metadata.packageName = resourceTable.currentResPackage.name
-                metadata.packageVersion = resourceTable.versionInfo.versionName
-                metadata.metaInfo.versionInfo = resourceTable.versionInfo
-                metadata.metaInfo.sdkInfo = resourceTable.sdkInfo
-            }
-        } finally {
-            extInputFile.close()
-        }
-    }
-
-    /**
-     * Execute patches added the patcher.
+     * Execute the patcher.
      *
      * @param stopOnError If true, the patches will stop on the first error.
-     * @return A pair of the name of the [Patch] and its [PatchResult].
+     * @return A pair of the name of the [Patch] and a [PatchException] if it failed.
      */
-    fun executePatches(stopOnError: Boolean = false): Sequence<Pair<String, Result<PatchResultSuccess>>> {
+    fun execute(stopOnError: Boolean = false) = sequence {
         /**
          * Execute a [Patch] and its dependencies recursively.
          *
          * @param patchClass The [Patch] to execute.
          * @param executedPatches A map of [Patch]es paired to a boolean indicating their success, to prevent infinite recursion.
-         * @return The result of executing the [Patch].
          */
         fun executePatch(
-            patchClass: Class<out Patch<Context>>,
-            executedPatches: LinkedHashMap<String, ExecutedPatch>
-        ): PatchResult {
+            patchClass: PatchClass,
+            executedPatches: HashMap<String, ExecutedPatch>
+        ) {
             val patchName = patchClass.patchName
 
-            // if the patch has already applied silently skip it
+            // If the patch has already executed silently skip it.
             if (executedPatches.contains(patchName)) {
                 if (!executedPatches[patchName]!!.success)
-                    return PatchResultError("'$patchName' did not succeed previously")
+                    throw PatchException("'$patchName' did not succeed previously")
```

Comment:
This does not give any consumer the chance to handle `PatchException`s properly.
How about `PatchException` becomes a sealed class with categorized exception subclasses, which only contain useful information? One example would be `PatchException.DidNotSucceed(patchName)`, where any consumer can handle this to their liking and possibly translate these errors.

Reply:
In another review something similar has been suggested already. Patches don't need to return anything and can throw an exception such as PatchException(message) that the patcher can handle. Further abstractions are not necessary and can be done by patches themselves.

Reply:
I forgot that patches return `PatchException`. How about we do the proposal I said, but then call that `PatcherException`? Then you could have `PatcherException.Patch(patchException)`.

Reply:
PatcherException should be thrown by the patcher and not patches. Patches should throw PatchException the way I explained it above.

Reply:
@oSumAtrIX That's literally what I said?? I just added that the patcher should catch and then ***wrap*** them, so that `save()` only ever throws `PatcherException`.

Reply:
Instead of `PatcherException.Patch(patchException)` it should be `PatcherException(patchException)`. This level of abstraction is not necessary.

Reply:
Then how about the other categorized exceptions I mentioned, or do you mean just the patch exception one? This whole thing is about the fact that the consumers can handle exceptions programmatically instead of having to do error-prone `exception.toString().contains("blah")`.

Reply:
`patcherException.exception is PatchException` suffices.

Reply:
That's not what I said, at all 🤦‍♂️ Please respond to what I *actually* said.

Reply:
I responded to what you said. You said, "instead of having to do error-prone exception.toString().contains("blah")." to which I responded what suffices instead of what you said. Please remain formal. Comments such as "🤦‍♂️" are exactly the wrong way to do formal reviews. As I explained earlier, no other categories are required here. The patches can simply throw a `PatchException` class and the patcher can wrap it in a `PatcherException`.

Reply:
I am still not sure what to do here...

My question is, does anyone (cli/manager) actually need more context other than the message?

Reply:
@Sculas You can check the type of exception that `PatcherException` wraps as explained above.

Reply:
@Axelen123 A `PatcherException` that wraps `PatchException` can be propagated to the caller, though that would break any loops such as that of executing patches.

Reply:
Alright